src/llm/wmt_data_utils.py

```diff
@@ -0,0 +1,243 @@
+
+    Returns:
+        Union[str, list[str]]: The decoded tokens.
+    """
+    # TODO: infer batch from the shape of `tokens`
+    if batch:
+        return tokenizer.decode_batch(tokens)
+    else:
+        return tokenizer.decode(tokens)
+
+
+def _collate_fn(batch, source, target, tokenizers):
+    """
+    Collate function for the dataloader. This will collate the batches into a single tensor.
+
+    Args:
+        batch (list): The batch of data.
+        source (str): The source language.
+        target (str): The target language.
+        tokenizers (dict): The tokenizers for the specified dataset name.
+
+    Returns:
+        tuple: The collated batches.
+    """
+    zipped_list = [
+        (
+            sample["translation"][source],
+            sample["translation"][target],
+        )
+        for sample in batch
+    ]
+    source_batch, target_batch = list(zip(*zipped_list))
+
+    # enable padding for the tokenizers
+    # we do the longest sequence length for padding
+    tokenizers[source].enable_padding(pad_id=1, pad_token="<pad>")
+    tokenizers[target].enable_padding(pad_id=1, pad_token="<pad>")
+
+    # tokenize the batches
+    tokenized_source_batch = [token.ids for token in tokenizers[source].encode_batch(source_batch)]
+    tokenized_target_batch = [token.ids for token in tokenizers[target].encode_batch(target_batch)]
+
+    # torch tensors
+    tokenized_source_batch = torch.tensor(tokenized_source_batch)
+    tokenized_target_batch = torch.tensor(tokenized_target_batch)
+
+    return tokenized_source_batch, tokenized_target_batch
+
+
+def get_wmt_dataloader(
+    dataset,
+    batch_size,
+    source_lang,
+    target_lang,
+    tokenizer_path="data",
+):
+    """
+    Get the dataloader for the specified dataset name.
+
+    Args:
+        dataset (str): The name of the dataset.
+        batch_size (int): The batch size.
+        source_lang (str): The source language.
+        target_lang (str): The target language.
+        tokenizer_path (str): The path to the tokenizers.
+
+    Returns:
+        DataLoader: The dataloader for the specified dataset name.
+    """
+    tokenizers = get_wmt_tokenizers(f"{source_lang}-{target_lang}", tokenizer_path)
+    return DataLoader(
+        dataset=dataset,
+        batch_size=batch_size,
+        collate_fn=partial(
+            _collate_fn, source=source_lang, target=target_lang, tokenizers=tokenizers
+        ),
+    )
+
+
+def create_separate_wmt_tokenizers(
+    vocab_size: int = 30000,
+    min_frequency: int = 2,
+    special_tokens: list[str] = ["<s>", "<pad>", "</s>", "<unk>", "<sep>"],
+    batch_size: int = 1000,
+    dataset_name: str = "de-en",
+):
+    """
+    Create and train two separate tokenizers for Source and Target languages from the WMT dataset.
+
+    Args:
+        vocab_size (int): Total vocabulary size for the tokenizers. If vocab size is 30000, english and german tokenizers will have 15000 tokens each.
```

Comment:
english -> source
german -> target

Maybe we should have different vocal sizes, and the user should take a call here. WDYT?

Reply:
Totally okay with the current setup (I was just thinking how would that be)

We can move this to our next iteration as well.